### PR TITLE
Update docs with slot predicate validation

### PR DESCRIPTION
**Pull request by ancalita**
**Proposed changes**:
- ...

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [x] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: docs/docs/concepts/flows.mdx [hunk 1]
```diff
@@ -483,6 +483,44 @@ flow is completed.
   scope: "flow"
 ```
 
+#### Slot value validation
```
Comment (m-vdb) [inline docs/docs/concepts/flows.mdx:486]: I think we refer to this concept like this throughout the docs
```suggestion
#### Slot validation
```

File: docs/docs/concepts/flows.mdx [hunk 2]
```diff
@@ -483,6 +483,44 @@ flow is completed.
...
+You can now write slot value validation directly in your flow yaml file without the need to use custom actions.
+Add a `rejections` property to your `collect_information` step to define the validation rules.
+The `rejections` section is a list of mappings, where each mapping has an `if` and an `utter` property:
+- the `if` property is a condition written in natural language that is evaluated using the
+[pypred](https://github.com/armon/pypred) library.
+- the `utter` property is the name of the response the assistant will send if the condition evaluates to `True`.
+
+Here is an example:
+
+```yaml
```
Comment (m-vdb) [inline docs/docs/concepts/flows.mdx:497]:
```suggestion
```rasa-yaml title="flows.yml"
```

File: docs/docs/concepts/flows.mdx [hunk 3]
```diff
@@ -483,6 +483,44 @@ flow is completed.
...
+    description: This flow verifies if the user is eligible for creating an account.
+    name: verify eligibility
+    steps:
+      - id: "ask_age"
+        collect: "age"
+        rejections:
+          - if: age < 1
+            utter: utter_invalid_age
+          - if: age < 13
+            utter: utter_age_at_least_13
```
Comment (m-vdb) [inline docs/docs/concepts/flows.mdx:509]: nit: maybe something like 18 would be more realistic

File: docs/docs/concepts/flows.mdx [hunk 4]
```diff
@@ -483,6 +483,44 @@ flow is completed.
   scope: "flow"
 ```
 
+#### Slot value validation
+
+You can now write slot value validation directly in your flow yaml file without the need to use custom actions.
+Add a `rejections` property to your `collect_information` step to define the validation rules.
```
Comment (m-vdb) [inline docs/docs/concepts/flows.mdx:489]: I think the first sentence is written from the POV of an existing rasa user; instead we should write it from the POV of a new user that doesn't know about custom actions. I don't think it adds much to mention actions here (and could confuse people).
```suggestion
You can define slot validation rules directly in your flow yaml file by adding a `rejections` property under
any `collect_information` step.
```

File: docs/docs/concepts/flows.mdx [hunk 5]
```diff
@@ -483,6 +483,44 @@ flow is completed.
...
+The `rejections` section is a list of mappings, where each mapping has an `if` and an `utter` property:
```
Comment (m-vdb) [inline docs/docs/concepts/flows.mdx:490]: if they're both required I would mention it here

File: docs/docs/concepts/flows.mdx [hunk 6]
```diff
@@ -483,6 +483,44 @@ flow is completed.
...
+- the `if` property is a condition written in natural language that is evaluated using the
+[pypred](https://github.com/armon/pypred) library.
+- the `utter` property is the name of the response the assistant will send if the condition evaluates to `True`.
```
Comment (m-vdb) [inline docs/docs/concepts/flows.mdx:493]:
```suggestion
- the `utter` property is the name of the [response](./responses.mdx) the assistant will send if the condition evaluates to `True`.
```

File: docs/docs/concepts/flows.mdx [hunk 7]
```diff
@@ -483,6 +483,44 @@ flow is completed.
...
+        collect: "age"
+        rejections:
+          - if: age < 1
+            utter: utter_invalid_age
+          - if: age < 13
+            utter: utter_age_at_least_13
+        next: "ask_email"
+```
+
+Upon every validation failure, the assistant will again make another attempt to collect the information.
```
Comment (m-vdb) [inline docs/docs/concepts/flows.mdx:513]:
```suggestion
Upon every validation failure, the assistant will automatically make another attempt to collect the information.
```

File: docs/docs/concepts/flows.mdx [hunk 8]
```diff
@@ -483,6 +483,44 @@ flow is completed.
...
+This will be repeated until none of the rejection conditions pass. Once all rejections have evaluated to `False`, the
```
Comment (m-vdb) [inline docs/docs/concepts/flows.mdx:514]: I am not sure what you mean by `this` here. Can you explain what is repeated exactly?

File: docs/docs/concepts/flows.mdx [hunk 9]
```diff
@@ -483,6 +483,44 @@ flow is completed.
...
+assistant moves on to the `next` step with the collected value.
+
+If the predicate defined in the `if` property is invalid or missing, the assistant will log an error and respond with the
```
Comment (m-vdb) [inline docs/docs/concepts/flows.mdx:517]: I'm a bit confused here. For me there are 3 error cases:
- `if` property missing -> should be detected in `rasa data validate` + runtime error as failsafe using `utter_internal_error_rasa`
- `if` property defines an invalid predicate -> should be detected in `rasa data validate` + runtime error as failsafe using `utter_internal_error_rasa`
- the execution of the predicates raises an error -> runtime error that uses `utter_internal_error_rasa`


Does it work this way?

File: docs/docs/concepts/flows.mdx [hunk 10]
```diff
@@ -483,6 +483,44 @@ flow is completed.
...
+`utter_internal_error_rasa` response which by default is
+`Sorry, I'm having trouble understanding you right now. Please try again later.`. However, you can customize this in
+your domain file. We recommend using end-to-end testing to ensure that your flows work as expected.
```
Comment (m-vdb) [inline docs/docs/concepts/flows.mdx:520]: usually we give an example of how this customisation looks like. Wondering if it makes sense here or if it makes the page look too complicated. Thoughts?

File: docs/docs/concepts/flows.mdx (hunk 8 above)
Comment (ancalita) [reply docs/docs/concepts/flows.mdx:514]: The bot re-asking the `utter_ask` response in order to collect the information. I'll make it more explicit.

File: docs/docs/concepts/flows.mdx (hunk 9 above)
Comment (ancalita) [reply docs/docs/concepts/flows.mdx:517]: > `if` property missing

I meant if it's explicitly coded as `null` / None in the yaml file. The schema for the yaml format should catch an `if` required property that is missing.
The error cases you listed are correct.

File: docs/docs/concepts/flows.mdx (hunk 10 above)
Comment (ancalita) [reply docs/docs/concepts/flows.mdx:520]: Hmm, tbh I should try it out first as well, haven't yet tested adding this response in the domain file with a different text. Not sure what happens when the default flows are merged (that's where this response resides in OSS).

Comment (ancalita) [reply docs/docs/concepts/flows.mdx]: Confirming that I tested the customization of the `utter_internal_error_rasa` and it worked fine, the overridden new response is displayed in the chat. There's a UserWarning displayed during training that duplicate responses were found for `utter_internal_error_rasa`, but training completes successfully.
I think adding an example for overriding the response would clutter the section, i can just say add a response `utter_internal_error_rasa` to your domain file with your custom text.

**Review by m-vdb — APPROVED** [review]
🚀 thanks Anca!

**Pull request merged**